### ISSUE TITLE
DOM rendering as a subscription?

### ISSUE THREAD

**Issue by sergey-shpak** [jorgebucaran/hyperapp#766]
Fasten seat belts, here we go again )) (inspired by lifecycle discussion at [#717](https://github.com/jorgebucaran/hyperapp/issues/717)) ...

What is the difference between fetching data and manipulating DOM?
Hyperapp doesn't fetch any data for you, so why it should manipulate DOM for you?

Moreover, let's assume I want to `render` state to canvas or `transform` it to xslt,
should this be builtin to hyperapp core? I don't think so, but why do we do it for html?   

Talking about state-driven apps, hyperapp V2 should not control anything except the state.
Fetching data, either Manipulating DOM is communication with external world, 
and should be done with interop mechanics provided by Hyperapp V2

And guess what? It's already possible with Hyperapp V2
```javascript
import { app, h } from 'hyperapp'
import Dom from 'subscriptions/dom'

// Dom.patch - is a simple subs wrapper for `superfine`
// Dom.on - simple eventbus for dom events

const ready = (state, params) =>
  ({ ...state, render: params.status })

// and yes, `view` is an action
const view = state => <div>view</div>

app({
  init: { render: false },
  subscriptions: state => [
    state.render &&
      Dom.patch({
        action: view, 
        container: document.body
      }),
    Dom.on({ ready })
  ],
  // container is not required anymore
  container: document.createElement('fake')
})
```
_______
### The difference
The difference is not what to implement, the difference is how to implement:
- core becomes more pure (it doesn't mutate any external source) 
- more freedom to manage app `rendering` (change lib, algo, etc)
- more control over when and where to `render` app

At the end it looks possible to simplify app initialization to simple subs/interop interface:
```javascript
// Dispatch and Dom are subscriptions
const init = () => ({ render: false })
const ready = state => ({ ...state, render: true })

app(state => [
  !state && Dispatch(init),
  state.render && Dom.patch(view, document.body),
  Dom.on({ ready })
])
```
Any thoughts?

**Comment by jorgebucaran** [comment]
Feel free to discuss this, but the VDOM is not something I want to split from Hyperapp.

**Comment by jorgebucaran** [comment]
@sergey-shpak Sorry, no. It is not the direction I want to take with Hyperapp. I don't want to contribute another Swiss Army Knife UI framework to JavaScript. Too flexible, one-size-fits-it-all, one abstraction after another. 

Some abstraction is good. V2 will be more abstract than V1, but not too much as to scare people away. I see the hidden elegance in exposing the DOM through a subscription, but DOM manipulation is the scope I've defined for Hyperapp, so I want it to be built-in, no-plumbing. 

If you can work around this constraint and use subscriptions for another kind of DOM manipulation, .e.g., observing elements, thus removing the need for lifecycle events, I'm in. 💯

**Comment by mrozbarry** [comment]
I think you've suggested something that's really neat, but I think very much out of scope of what hyperapp is. Also, you definitely wouldn't want rendering as a subscription. Subscriptions are to get outside events into the hyperapp loop, and not really the other way around (unless I missed something about the subscription implementation).

To address out of scope of hyperapp, at the very top of the README:

> **Standalone** — Do more with less. Hyperapp combines state management with a virtual DOM engine that supports keyed updates & lifecycle events — all with no dependencies.

Because hyperapp is pre-combined the state management and virtual dom rendering, the code can make certain efficiencies, in terms of waiting on renders, reducing the call stack, etc. If the dom implementation were decoupled, we'd definitely lose some speed, and I'm not sure about size.

Anyway, I think you have a cool idea, it's just too far removed from the current hyperapp. I'd be interested if you did something neat with redux/mobx/some-functional-state-library with superfine that implemented the interfaces you describe, but that's not really a discussion for a hyperapp issue at that point.

**Comment by jorgebucaran** [comment]
@sergey-shpak I do like the idea of using subscriptions with mutation observers like shown in your pen and I think this is how we are going to get rid of lifecycle events! 🎉 

cc @zaceno @SkaterDad  @vladimirkosmala @atomiks @dmitrykurmanov 🙏

**Comment by SkaterDad** [comment]
@jorgebucaran I haven't had any time to look into this, and have about 0 experience with MutationObservers.

Could someone with experience set up an example which mounts a 3rd party UI widget that can receive props & call actions?  Or is that not the intention here?

**Comment by jorgebucaran** [comment]
@sergey-shpak Is there anything else you'd like to share with us? I think we're good to close here. There's nothing actionable, and lifecycle events are out of V2 as well.
